posts/2022-12-15-Rust-1.66.0.md

```diff
@@ -0,0 +1,171 @@
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+        v.push(i);
+    }
+}
+
+pub fn bench_push() -> Duration { 
+    let mut v = Vec::with_capacity(4);
+    let now = Instant::now();
+    push_cap(&mut v);
+    now.elapsed()
+}
+```
+
+If you inspect the optimized output of the compiler on x86_64, you'll notice that it looks rather short:
+
+```asm
+example::bench_push:
+  sub rsp, 24
+  call qword ptr [rip + std::time::Instant::now@GOTPCREL]
+  lea rdi, [rsp + 8]
+  mov qword ptr [rsp + 8], rax
+  mov dword ptr [rsp + 16], edx
+  call qword ptr [rip + std::time::Instant::elapsed@GOTPCREL]
+  add rsp, 24
+  ret
+```
+
+In fact, the entire function `push_cap` we wanted to benchmark has been optimized away!
+
+We can work around this using the newly stabilized `black_box` function. Functionally, `black_box` is not very interesting: it takes the value you pass it and passes it right back. Internally, however, the compiler treats `black_box` as a function that could do anything with its input and return any value (as its name implies).
+
+This is very useful for disabling optimizations like the one we see above. For example, we can hint to the compiler that the vector will actually be used for something after every iteration of the for loop.
+
+```rust
+use std::hint::black_box;
+
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+        v.push(i);
+        black_box(v.as_ptr());
+    }
+}
+```
+
+Now we can find the unrolled for loop in our [optimized assembly output](https://rust.godbolt.org/z/Ws1GGbY6Y):
+
+```asm
+  mov dword ptr [rbx], 0
+  mov qword ptr [rsp + 8], rbx
+  mov dword ptr [rbx + 4], 1
+  mov qword ptr [rsp + 8], rbx
+  mov dword ptr [rbx + 8], 2
+  mov qword ptr [rsp + 8], rbx
+  mov dword ptr [rbx + 12], 3
+  mov qword ptr [rsp + 8], rbx
+```
+
+You can also see a side effect of calling `black_box` in this assembly output. The instruction `mov qword ptr [rsp + 8], rbx` is uselessly repeated after every iteration. This instruction writes the address `v.as_ptr()` as the first argument of the function, which is never actually called.
+
+Notice that the generated code is not at all concerned with the possibility of allocations introduced by the `push` call. This is because the compiler is still using the fact that we called `Vec::with_capacity(4)` in the `bench_push` function. You can play around with the placement of `black_box`, or try using it in multiple places, to see its effects on compiler optimizations.
+
+### cargo remove
+
+In Rust 1.62.0 we introduced `cargo add`, a command line utility to add dependencies to your project. Now you can use `cargo remove` to remove dependencies.
+
+### Stabilized APIs
+
+- [`proc_macro::Span::source_text`](https://doc.rust-lang.org/stable/proc_macro/struct.Span.html#method.source_text)
+- [`uX::{checked_add_signed, overflowing_add_signed, saturating_add_signed, wrapping_add_signed}`](https://doc.rust-lang.org/stable/std/primitive.u8.html#method.checked_add_signed)
```

Comment:
I do find the naming a bit unfortunate because there is a crate called [`uX`](https://crates.io/crates/ux) providing non-power-of-2 integers and someone might think that Rust has gotten that long thought after ability now... it certainly got me excited for a minute.